### PR TITLE
feat(pricing): Add taxable did & resource messages, supply deflation, block rewards & `simapp` tests

### PR DESCRIPTION
**Pull request by Eengineer1**
# Introducing Taxable Messages

```mermaid
graph TB
a[Tx] --> b{CheckTx / ReCheckTx} 
b --> c{IsTaxableTx Lite Validation} 
c --> |True| d[No-Op, will be handled by PostHandler]
c --> |False| e[AnteHandler DeductFeesDecorator]
a --> f{DeliverTx}
f --> g[MsgServer Dynamic Validation]
g --> h[PostHandler TaxDecorator]
h --> i{IsTaxableTx Lite Validation}
i --> |Yes| j[DeductTax]
j --> k[Burn Portion]
j --> l[Block Rewards Portion]
i --> |No| m[No-Op, already handled by AnteHandler]
```
## Scopes
1. Supply deflation
2. Increase block rewards

## Relationships
1. MsgServer dynamic validation fails, PostHandler fails as well, state is not committed, block gas is consumed, tax is not deducted.
2. PostHandler fails, state is not committed, block gas is consumed, tax is not deducted.

     e.g. 
     - Insufficient fee payer funds.
     - Asked for a non-existing or revoked fee grant.

### CODE REVIEW

**Comment by ankurdotb** [comment]
Task linked: [DEV-1599 v1 Identity Pricing](https://app.clickup.com/t/6600954/DEV-1599)

File: ante/fee.go [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+			return sdkerrors.ErrInvalidRequest.Wrap("fee grants are not enabled")
+		} else if !feeGranter.Equals(feePayer) {
+			err := dfd.feegrantKeeper.UseGrantedFees(ctx, feeGranter, feePayer, fee, sdkTx.GetMsgs())
+			if err != nil {
+				return sdkerrors.Wrapf(err, "%s does not not allow to pay fees for %s", feeGranter, feePayer)
+			}
+		}
+
+		deductFeesFrom = feeGranter
+	}
```
Comment (ankurdotb) [inline ante/fee.go:126]: Do we currently or in the future plan on using `feegrant` module?

File: ante/fee_test.go [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+	antehandler := sdk.ChainAnteDecorators(dfd)
+
+	_, err = antehandler(s.ctx, tx, false)
+
+	s.Require().NotNil(err, "Tx did not error when fee payer had insufficient funds")
+
+	// Set account with sufficient funds
+	s.app.AccountKeeper.SetAccount(s.ctx, acc)
+	err = testutil.FundAccount(s.app.BankKeeper, s.ctx, addr1, sdk.NewCoins(sdk.NewCoin("cheq", sdk.NewInt(200))))
+	s.Require().NoError(err)
```
Comment (ankurdotb) [inline ante/fee_test.go:148]: People always provide their fees in base denom in ncheq, so isn't this supposed to be in ncheq>

Comment (Eengineer1) [reply ante/fee_test.go]: These are native tests in `u` units, so the denom is not relevant as long as the comparison is under the same metric, but for sure it makes sense to keep in the correct line.

The following three are ensuring we're following default fee deduction logic for native SDK tx types:

- `TestDeductFeeDecorator_ZeroGas`
- `TestEnsureMempoolFees`
- `TestDeductFees`

The updated version accounts for the denominations across the whole logic.

I've pushed a

File: ante/fee.go (hunk 1 above)
Comment (Eengineer1) [reply ante/fee.go:136]: Embedded by default in fee-checking closures, for upstream compatibility.
As for the module, `feegrant` enabled by default, so easier to keep than to ditch in terms of effort needed.

File: app/app.go [hunk 3]
```diff
@@ -680,17 +703,34 @@ func New(
...
 				}
+			} else {
+				println("Version map already initialized")
+
+				// Add defaults for staking subspace
+				stakingSubspace, _ := app.ParamsKeeper.GetSubspace(stakingtypes.ModuleName)
+				stakingSubspace.Set(ctx, stakingtypes.KeyMinCommissionRate, sdk.NewDec(0))
+
+				// Add defaults for resource subspace
+				resourceSubspace := app.GetSubspace(resourcetypes.ModuleName)
```
Comment (askolesov) [inline app/app.go:716]: Shouldn't we do the same for the did module?

Comment (Eengineer1) [reply app/app.go:716]: We can add it for safety.
Per our discussion, I was more worried on skipping InitGenesis for resource mostly.

File: app/app.go [hunk 4]
```diff
@@ -762,11 +802,11 @@ func (app *App) InitChainer(ctx sdk.Context, req abci.RequestInitChain) abci.Res
 	// Now this would break consensus with existing networks.
 	// so ModuleVersionMap is initialized as part of upgrade xxx.
 	// app.UpgradeKeeper.SetModuleVersionMap(ctx, app.mm.GetVersionMap())
-
 	var genesisState GenesisState
 	if err := tmjson.Unmarshal(req.AppStateBytes, &genesisState); err != nil {
 		panic(err)
 	}
+	// app.UpgradeKeeper.SetModuleVersionMap(ctx, app.mm.GetVersionMap())
```
[the remark this answers is not in the document]
Comment (Eengineer1) [reply app/app.go:809]: Yeah ... probably better to keep the commented line where it's supposed to be (not 804), imo.

**Comment by askolesov** [comment]
We think it's ok to merge it as is now to unblock release process and continue detailed review tomorrow.

**Review by askolesov — APPROVED** [review]

**Pull request merged**